mmengine/optim/optimizer/default_constructor.py

```diff
@@ -297,8 +298,69 @@ def __call__(self, model: nn.Module) -> OptimWrapper:
             # set param-wise lr and weight decay recursively
             params: List = []
             self.add_params(params, model)
-            optimizer_cfg['params'] = params
+            # grouping parameters with the same hyper-parameters
+            if self.paramwise_cfg.get('reduce_param_groups', True):
+                optimizer_cfg['params'] = reduce_param_groups(params)
+            else:
+                optimizer_cfg['params'] = params
+            # enable foreach for pytorch 1.12.0+ to speed up training
+            if digit_version(TORCH_VERSION) >= digit_version('1.12.0'):
+                optimizer_cfg.setdefault('foreach', True)
+            else:
+                optimizer_cfg.pop('foreach', None)
+
             optimizer = OPTIMIZERS.build(optimizer_cfg)
         optim_wrapper = OPTIM_WRAPPERS.build(
             optim_wrapper_cfg, default_args=dict(optimizer=optimizer))
         return optim_wrapper
+
+
+def _expand_param_groups(params: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
+    """Transform parameter groups into per-parameter structure. Later items in
+    `params` can overwrite parameters set in previous items.
+
+    Ref: https://github.com/facebookresearch/detectron2/blob/main/detectron2/solver/build.py
+
+    Args:
+        params (List[Dict[str, Any]]): The parameter groups.
+
+    Returns:
+        List[Dict[str, Any]]: List of expanded parameter groups.
+    """  # noqa: E501
+    ret: dict = defaultdict(dict)
+    for item in params:
+        assert 'params' in item
+        cur_params = {x: y for x, y in item.items() if x != 'params'}
+        for param in item['params']:
+            ret[param].update({'params': [param], **cur_params})
+    return list(ret.values())
+
+
+def reduce_param_groups(params: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
```

Comment:
Could we merge these two functions into one? just like this

```suggestion
groups = defaultdict(set)
for param_group in params:
    params_value = {x: y for x, y in param_group.items() if x != 'params'}
    for param in param_value:
        groups[param_group].add(param_value)
  ret = []
  ...
```